clients/src/main/java/org/apache/kafka/common/network/ConnectionRegistry.java

```diff
@@ -0,0 +1,258 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.network;
+
+import java.net.InetAddress;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.metrics.Gauge;
+import org.apache.kafka.common.metrics.MetricConfig;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.security.auth.KafkaPrincipal;
+import org.apache.kafka.common.security.auth.SecurityProtocol;
+
+/**
+ * Maintains metadata about each active connections and exposed various metrics about the connections
+ * and the clients.
```

Comment:
I don't think it's necessary to keep a central registry of all this information for all connections.  We really just need the metrics, most of which can just be simple counters.  If we need more information about a connection, we can look at the request context of that connection.  But it doesn't have to be stored here.

Reply:
Let me explain why I went down this path.

1) As you said, the request context can be used to get the information about the connection. The issue is that that information must come from some place. It could come directly from the KafkaChannel if SASL is used and an ApiVersionsRequest with the information is received during the SASL initialisation. Unfortunately, some clients, including the java one, does not provide the information during the SASL initialisation but in a second ApiVersionsRequest which is handled in the KafkaApis layer this time. This means that a place is required to store and update the name and the version.

2) KIP-511 has proposed to add a new `Gauge<List<Map<String, String>>` `kafka.server:type=ClientMetrics,name=Connections` which lists all the active connections with their metadata (client id, software name, software version, etc.). This requires a list of all the active connections to be maintained somewhere to back the gauge. We could argue that this could be replace by counters for each combination. I preferred the Gauge to limit the number of metrics.

I have considered the following alternative to the registry but did not choose it. You may prefer it.

* As the Selector already maintains a list of the active KafkaChannels (equivalent to connections), we could add all the require metadata as attributes in the KafkaChannel and use the selector as source of truth. This would require to add the clientId in channel for instance.

* To avoid having the pass the KafkaChannel up to the KafkaApis layer to update the name and the version when the ApiVersionsRequest is received, we could partially or completely move the handling of the AVR in the processor to update the KafkaChannel. The later would impact the throttling, the metrics and the request logs.

* As many selectors are run (one per processor if not mistaken), we could have counters per selector to limit the locking required to update them. For the Gauge which lists all the connections, we could also have one per Selector but it is less convenient so for this one, I would propose to have a way to list all the KafkaChannel across all the selectors.

The major advantage of this approach is that the "per selector/processor" does not require any locking.

The downsides:
* Moving the handling of the AVR in the processor has wide impact on the throttling, the metrics of the AVR and the request log which is based on the request send to the KafkaApis layer.
* It makes exposing the metadata via other API little harder. For instance, one could think of adding an Admin API to list all the active connections of a broker (already got such request from our colleagues).

I was not comfortable with moving the handling of the AVR in the processor when I looked at the options so I went with the current approach.

What do you think?

Reply:
Thanks for the explanation, and thanks for fixing the synchronization.  As an optimization, I think we should use something like `ConcurrentMap` here rather than synchronized blocks, so that we can minimize the amount of time threads spend waiting.  This will be a bit more tricky to use, but more scalable.

I guess when I thought about KIP-511, I thought of it in terms of metrics and perhaps occasional samples of connections (similar to how we do request sampling by logging a few selected requests).  I don't see why we'd ever need a full snapshot of all existing connections.  The snapshot would probably be out of date by the time it had been returned, since new connections are closed and opened all the time.

Reading the KIP more carefully, I see that KIP-511 does specify a metric which essentially requires each connection to register itself.  Considering we don't even have a way to visualize or graph this metric, I'm not sure this belongs in JMX.  I have to think about this more...